### PR TITLE
Issue16

### PR DESCRIPTION
**Pull request by johanix**

### CODE REVIEW

File: common/fsmops.go [hunk 1]
```diff
@@ -94,7 +102,14 @@ func (mdb *MusicDB) ZoneDetachFsm(dbzone *Zone, fsm, fsmsigner string) (error, s
...
 
-	mdb.mu.Lock()
+	localtx, tx, err := mdb.StartTransaction(tx)
+	if err != nil {
+		log.Printf("ZoneJoinGroup: Error from mdb.StartTransaction(): %v\n", err)
+		return err, "fail"
+	}
+	defer mdb.CloseTransaction(localtx, tx, err)
+
+	// mdb.mu.Lock()
```
[the remark this answers is not in the document]
Comment (lundqvistm) [reply common/fsmops.go:112]: And all the rest as well

**Review by romu42 — APPROVED** [review]

File: common/zoneops.go [hunk 2]
```diff
@@ -19,34 +19,51 @@ func (z *Zone) SignerGroup() *SignerGroup {
...
+	dbzone, _ := mdb.GetZone(tx, fqdn)
 	if dbzone.Exists {
 		return fmt.Errorf("Zone %s already present in MuSiC system.", fqdn), ""
 	}
 
+	const AZsql = `
+INSERT INTO zones(name, zonetype, state, statestamp, fsm, fsmmode)
+VALUES (?, ?, ?, datetime('now'), ?, ?)`
+
 	stmt, err := mdb.Prepare(AZsql)
```
Comment (lundqvistm) [inline common/zoneops.go:42]: Why  use a prepared statement here?

File: common/zoneops.go [hunk 3]
```diff
@@ -19,34 +19,51 @@ func (z *Zone) SignerGroup() *SignerGroup {
...
 
-	mdb.mu.Lock()
+	// mdb.mu.Lock()
+//	tx, err := mdb.Begin()
+//	if err != nil {
+//		log.Printf("AddZone: Error from mdb.Begin(): %v", err)
+//	}
+	// defer tx.Commit()
+	
 	_, err = stmt.Exec(fqdn, z.ZoneType, "", "", z.FSMMode)
```
Comment (lundqvistm) [inline common/zoneops.go:54]: The variables you leave empty don't have default values in the db?

File: common/zoneops.go [hunk 4]
```diff
@@ -311,7 +381,16 @@ SELECT name, zonetype, state, fsmmode, COALESCE(statestamp, datetime('now')) AS
...
 )
 
-func (mdb *MusicDB) GetZone(zonename string) (*Zone, bool) {
+func (mdb *MusicDB) GetZone(tx *sql.Tx, zonename string) (*Zone, bool) {
+
+	localtx, tx, err := mdb.StartTransaction(tx)
+	if err != nil {
+		log.Printf("ZoneJoinGroup: Error from mdb.StartTransaction(): %v\n", err)
+		// return err, "fail"
+		return nil, false // XXX: Want an err return also.
```
Comment (lundqvistm) [inline common/zoneops.go:390]: XXX == TODO?

File: musicd/apiserver.go [hunk 5]
```diff
@@ -282,10 +282,10 @@ func APIzone(conf *Config) func(w http.ResponseWriter, r *http.Request) {
 		case "step-fsm":
 			// var zones map[string]music.Zone
 			// var success bool
-			// err, resp.Msg, zones = mdb.ZoneStepFsm(dbzone, zp.FsmNextState)
+			// err, resp.Msg, zones = mdb.ZoneStepFsm(nil, dbzone, zp.FsmNextState)
```
Comment (lundqvistm) [inline musicd/apiserver.go:285]: Can these comments be removed instead?

File: musicd/apiserver.go [hunk 6]
```diff
@@ -422,50 +422,50 @@ func APIsigner(conf *Config) func(w http.ResponseWriter, r *http.Request) {
 			Client: r.RemoteAddr,
 		}
 
-		dbsigner, _ := mdb.GetSigner(&sp.Signer, false) // not apisafe
+		dbsigner, _ := mdb.GetSigner(nil, &sp.Signer, false) // not apisafe
```
Comment (lundqvistm) [inline musicd/apiserver.go:425]: Why is it not apisafe?

File: musicd/fsmengine.go [hunk 7]
```diff
@@ -194,18 +195,18 @@ func PushZones(conf *Config) error {
...
 	log.Printf("PushZones: will push on these zones: %v", zones)
 	for _, z := range zones {
-		PushZone(conf, z)
+		xxxPushZone(conf, z)
 	}
 	return nil
 }
 
-func PushZone(conf *Config, z string) error {
+func xxxPushZone(conf *Config, z string) error {
```
Comment (lundqvistm) [inline musicd/fsmengine.go:203]: Not the best function name..

**Pull request merged**